packages/core/graph/src/AdjacencyList.js

```diff
@@ -0,0 +1,1195 @@
+ */
+const NEXT_HASH: 3 = 3;
+/**
+ * The offset from an edge index at which the 'to' node's
+ * next incoming edge is stored.
+ */
+const NEXT_IN: 4 = 4;
+/**
+ * The offset from an edge index at which the 'to' node's
+ * previous incoming edge is stored.
+ */
+const PREV_IN: 5 = 5;
+/**
+ * The offset from an edge index at which the 'from' node's
+ * next outgoing edge is stored.
+ */
+const NEXT_OUT: 6 = 6;
+/**
+ * The offset from an edge index at which the 'from' node's
+ * previous outgoing edge is stored.
+ */
+const PREV_OUT: 7 = 7;
+
+/** The offset from a node index at which the first incoming edge is stored. */
+const FIRST_IN: 0 = 0;
+/** The offset from a node index at which the first outgoing edge is stored. */
+const FIRST_OUT: 1 = 1;
+/** The offset from a node index at which the last incoming edge is stored. */
+const LAST_IN: 2 = 2;
+/** The offset from a node index at which the last outgoing edge is stored. */
+const LAST_OUT: 3 = 3;
+
+/** The upper bound above which the edge capacity should be increased. */
+const LOAD_FACTOR = 0.7;
+/** The lower bound below which the edge capacity should be decreased. */
+const UNLOAD_FACTOR = 0.3;
+/** The max amount by which to grow the capacity of the edges array. */
+const MAX_GROW_FACTOR = 8;
+/** The min amount by which to grow the capacity of the edges array. */
+const MIN_GROW_FACTOR = 2;
+/** The amount by which to shrink the capacity of the edges array. */
+const SHRINK_FACTOR = 0.5;
+/** How many edges to accommodate in a hash bucket. */
+const BUCKET_SIZE = 2;
+
+/** The smallest functional node capacity. */
+const MIN_NODE_CAPACITY = 2;
+/** The largest possible node capacity. */
+const MAX_NODE_CAPACITY = Math.floor(
+  // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Errors/Invalid_array_length#what_went_wrong
+  (2 ** 31 - 1 - NODES_HEADER_SIZE) / NODE_SIZE,
+);
+
+/** The smallest functional edge capacity. */
+const MIN_EDGE_CAPACITY = 2;
+/** The largest possible edge capacity. */
+const MAX_EDGE_CAPACITY = Math.floor(
+  // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Errors/Invalid_array_length#what_went_wrong
+  (2 ** 31 - 1 - EDGES_HEADER_SIZE) / EDGE_SIZE / BUCKET_SIZE,
+);
+/** The size after which to grow the edge capacity by the minimum factor. */
+const PEAK_EDGE_CAPACITY = 2 ** 18;
+
+export const ALL_EDGE_TYPES: AllEdgeTypes = '@@all_edge_types';
+
+// eslint-disable-next-line no-unused-vars
+export type SerializedAdjacencyList<TEdgeType> = {|
+  nodes: Uint32Array,
+  edges: Uint32Array,
+|};
+
+// eslint-disable-next-line no-unused-vars
+export type AdjacencyListOptions<TEdgeType> = {|
+  edgeCapacity?: number,
+  nodeCapacity?: number,
+|};
+
+opaque type EdgeHash = number;
+
+/** The index of the edge in the edges array. */
+opaque type EdgeIndex = number;
+
+// From https://gist.github.com/badboy/6267743#32-bit-mix-functions
+function hash32shift(key: number): number {
```

Comment:
This crazy bit mixing function improves the uniformity of our hash table a fair amount without adding any noticeable time to hashing. We found this in the [v8 Map hash utils](https://github.com/nodejs/node/blob/238104c531219db05e3421521c305404ce0c0cce/deps/v8/src/utils/utils.h#L213)